apps/devhub/utils.py

```diff
@@ -10,92 +10,92 @@
+from validator.constants import SIGNING_SEVERITIES
 from validator.version import Version
 from . import tasks
 
 
-# FIXME: Move to `validator.constants`.
-SEVERITIES = {'trivial', 'low', 'medium', 'high'}
+def process_validation(validation, is_compatibility=False):
+    """Process validation results into the format expected by the web
+    frontend, including transforming certain fields into HTML,  mangling
+    compatibility messages, and limiting the number of messages displayed."""
 
+    if is_compatibility:
+        mangle_compatibility_messages(validation)
+
+    # Set an ending tier if we don't have one (which probably means
+    # we're dealing with mock validation results).
+    validation.setdefault('ending_tier', 0)
+
+    if not validation['ending_tier'] and validation['messages']:
+        validation['ending_tier'] = max(msg.get('tier', -1)
+                                        for msg in validation['messages'])
+
+    limit_validation_results(validation)
+
+    htmlify_validation(validation)
+
+    return validation
+
+
+def mangle_compatibility_messages(validation):
+    """Mangles compatibility messages so that the message type matches the
+    compatibility type, and alters totals as appropriate."""
 
-def make_validation_results(data, is_compatibility=False):
-    if data['validation']:
-        data['validation'] = limit_validation_results(escape_validation(
-            data['validation']))
-    data['error'] = hide_traceback(data['error'])
-    return data
+    compat = validation['compatibility_summary']
+    for k in ('errors', 'warnings', 'notices'):
+        validation[k] = compat[k]
 
+    for msg in validation['messages']:
+        if msg['compatibility_type']:
+            msg['type'] = msg['compatibility_type']
 
-def hide_traceback(error):
-    """Safe wrapper around JSON dict containing a validation result.
-    """
-    if not settings.EXPOSE_VALIDATOR_TRACEBACKS and error:
-        # Just expose the message, not the traceback
-        return error.strip().split('\n')[-1].strip()
-    else:
-        return error
 
+def limit_validation_results(validation):
+    """Limit the number of messages displayed in a set of validation results,
+    and if truncation has occurred, add a new message explaining so."""
 
-def limit_validation_results(validation, is_compatibility=False):
+    messages = validation['messages']
     lim = settings.VALIDATOR_MESSAGE_LIMIT
-    if lim:
-        del validation['messages'][lim:]
-        if validation.get('compatibility_summary'):
-            cs = validation['compatibility_summary']
-            compatibility_count = (
-                cs['errors'] + cs['warnings'] + cs['notices'])
+    if lim and len(messages) > lim:
+        # Sort messages by severity first so that the most important messages
+        # are the one we keep.
+        TYPES = {'error': 0, 'warning': 1, 'notice': 2}
+        messages.sort(key=lambda msg: TYPES.get(msg.get('type')))
```

Comment:
Nope. All of these functions modify the results in place. I considered starting with a `deepcopy`, but all of the callers currently use results that are fresh from `json.loads`, so it didn't seem worth it. This is why I changed most of the functions not to return the input value, though, since that makes the side-effects much more obvious.